### ISSUE TITLE
Conditionally emit the last message of a `StreamMessage`

### ISSUE THREAD

**Issue by ikhoon** [line/armeria#4816]
As a future work of #4727, users might want to dynamically emit the last message depending on whether the upstream publisher completes successfully or exceptionally.
```java
interface StreamMessage<T> {

   <T> StreamMessage<T> endWith(Function<@Nullable Throwable, ? extends T> finalizer) {
       // Conditionally appends an element using the finalizer
       return new SurroundingPublisher(null, this, finalizer);
   }
}
```

**Comment by injae-kim** [comment]
Please assign this to me~ I'll create PR right after #4727 is merged 🙇
